src/seismometer/data/pandas_helpers.py

```diff
@@ -153,11 +175,132 @@ def infer_label(dataframe: pd.DataFrame, label_col: str, time_col: str) -> pd.Da
+            right = right_filtered
+
+        max_lookback = pd.Timedelta(window_hrs, unit="hr") + min_offset  #Keep window the specified size
+        right = pd.merge(right, left[pks+[l_ref]], on=pks, how="left")
+        right = right[right[l_ref] <= right[r_ref]] #Filter to only events that happened at or after the prediction
+        right = right[right[l_ref] > (right[r_ref] - max_lookback)] #Filter to only events that happened within the window
+    
+    #Create a value counts dataframe where each event is a column containing the count of that event grouped by the primary keys
+    val_counts: pd.DataFrame = right.groupby(pks, as_index=False)[event_label].value_counts()
+    val_counts = val_counts.pivot(index=pks, columns=event_label, values='count').reset_index().fillna(0).rename(columns=event_name_map)
+
+    left = pd.merge(left, val_counts, on=pks, how="left") #Merge counts into left frame
+    count_cols = list(event_name_map.values())
+    left[count_cols] = left[count_cols].fillna(0) #Fill any missing counts for rows that didn't have any events
+    
+    return left
+
+
+def _merge_with_strategy(
+    predictions: pd.DataFrame,
+    one_event: pd.DataFrame,
+    pks: list[str],
+    *,
+    pred_ref: str = "Time",
+    event_ref: str = "Time",
+    event_display: str = "an event",
+    merge_strategy: str = "forward",
+) -> pd.DataFrame:
+    """
+    Merges the right frame into the left based on a set of exact match primary keys and merge strategy. 
+
+    Parameters
+    ----------
+    predictions : pd.DataFrame
+        The left frame, usually of predictions. Assumed to be sorted by time.
+    one_event : pd.DataFrame
+        The right frame to merge, assumed to be of events. Assumed to be sorted by time if applicable.
+    pks : list[str]
+        The list of columns to require exact matches during the merge.
+    pred_ref : str, optional
+        The column in the left (prediction) frame to use as a reference point in the distance match, by default 'Time'.
+    event_ref : str, optional
+        The column in the right (event) frame to use reference in the distance match, by default 'Time'.
+    event_display : str, optional
+        The name of the event to display in warning messages, by default "an event"
+    merge_strategy : str
+        The method to use when merging the event data, by default 'forward'.
+
+    Returns
+    -------
+    pd.DataFrame
+        The merged dataframe.
+    """
+    try:
+        ct_times = one_event[event_ref].notna().sum()
+
+        #If there are no times in the event frame, merge the first row for each group
+        if ct_times == 0:
+                #Set the filtered frame to the first row for each group and throw a value error which is passed before merging.
+                one_event_filtered = one_event.groupby(pks).first()
+                raise ValueError(f"No times found for {event_display}, merging first row for each group.")
```

Comment:
I can do a custom one, I just wasn't sure if y'all would be in favor of customer errors.

I could also potentially use [pandas.errors.EmptyDataError](https://pandas.pydata.org/docs/reference/api/pandas.errors.EmptyDataError.html) to avoid some of the work. That's used by pandas when reading empty CSVs but it seems decently descriptive here too.